src/flag_gems/ops/masked_select.py

```diff
@@ -0,0 +1,68 @@
+import logging
+
+import torch
+import triton
+import triton.language as tl
+
+from flag_gems.utils import broadcastable_to, libentry
+
+
+def cfggen():
+    configs = [
+        triton.Config({"BLOCK_SIZE": bs}, num_warps=w)
+        for w in [4, 8, 16, 32]
+        for bs in [256, 512, 1024, 2048, 4096]
+    ]
+    return configs
+
+
+@libentry()
+@triton.autotune(configs=cfggen(), key=["n_elements"])
+@triton.jit
+def masked_select_kernel(
+    inp_ptr,
+    select_mask_ptr,
+    prefix_sum_ptr,
+    out_ptr,
+    n_elements,
+    BLOCK_SIZE: tl.constexpr,
+):
+    pid = tl.program_id(axis=0)
+    offsets = pid * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
+    mask = offsets < n_elements
+
+    inp = tl.load(inp_ptr + offsets, mask=mask, other=0.0)
+    select_mask = tl.load(select_mask_ptr + offsets, mask=mask, other=0.0).to(tl.int1)
+    out_offset = tl.load(prefix_sum_ptr + offsets, mask=mask, other=0.0) - 1
+
+    tl.store(out_ptr + out_offset, inp, mask=(select_mask and mask))
+
+
+def masked_select(inp, mask):
+    logging.debug("GEMS MASKED SELECT")
+
+    inp_shape = tuple(inp.shape)
+    mask_shape = tuple(mask.shape)
+
+    if broadcastable_to(mask_shape, inp_shape):
+        mask = mask.expand(inp.shape)
+    elif broadcastable_to(inp_shape, mask_shape):
+        inp = inp.expand(mask.shape)
+    else:
+        raise RuntimeError(
+            "The shapes of the `mask` and the `input` tensor must be broadcastable"
+        )
```

Comment:
input and mask may not be broadcastable to each other's shape, but they may be broadcastable.

```text
In [2]: x = torch.arange(6).reshape(1,6)

In [3]: x
Out[3]: tensor([[0, 1, 2, 3, 4, 5]])

In [4]: index = torch.tensor([[True], [False], [True]])

In [5]: torch.masked_select(x, index)
Out[5]: tensor([0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5])

In [6]: torch.masked_select(x, index).shape
Out[6]: torch.Size([12])
```

Reply:
actually `(1, 6)` and `(3, 1)` should be broadcastable to (3, 6).
I will fix function `broadcatable_to` to make it right.